server/core/redis/redis.go

```diff
@@ -66,49 +66,136 @@ func NewWithClient(client *redis.Client, bucket string, globalBucket string) (*R
+			return false, currLock, enqueueStatus, errors.Wrap(err, "failed to deserialize current lock")
 		}
-		return false, currLock, nil
+		// checking if current lock is with the same PR
+		if currLock.Pull.Num == newLock.Pull.Num {
+			return false, currLock, enqueueStatus, nil
+		}
+
+		queueKey := r.queueKey(newLock.Project, newLock.Workspace)
+		currQueueSerialized, err := r.client.Get(ctx, queueKey).Result()
+		var queue models.ProjectLockQueue
+		if err == redis.Nil {
+			queue = models.ProjectLockQueue{}
+		} else if err != nil {
+			return false, currLock, enqueueStatus, errors.Wrap(err, "db transaction failed")
+		} else {
+			if err := json.Unmarshal([]byte(currQueueSerialized), &queue); err != nil {
+				return false, currLock, enqueueStatus, errors.Wrap(err, "failed to deserialize current queue")
+			}
+		}
+		// Lock is already in the queue
+		if indexInQueue := queue.FindPullRequest(newLock.Pull.Num); indexInQueue > -1 {
+			enqueueStatus = models.EnqueueStatus{
+				Status:     models.AlreadyInTheQueue,
+				QueueDepth: indexInQueue + 1,
+			}
+			return false, currLock, enqueueStatus, nil
+		}
+
+		// Not in the queue, add it
+		newQueue := append(queue, newLock)
+		newQueueSerialized, err := json.Marshal(newQueue)
+		if err != nil {
+			return false, currLock, enqueueStatus, errors.Wrap(err, "serializing")
+		}
+		err = r.client.Set(ctx, queueKey, newQueueSerialized, 0).Err()
+		if err != nil {
+			return false, currLock, enqueueStatus, errors.Wrap(err, "db transaction failed")
+		}
+		enqueueStatus = models.EnqueueStatus{
+			Status:     models.Enqueued,
+			QueueDepth: len(newQueue),
+		}
+		return false, currLock, enqueueStatus, nil
 	}
 }
 
 // Unlock attempts to unlock the project and workspace.
 // If there is no lock, then it will return a nil pointer.
 // If there is a lock, then it will delete it, and then return a pointer
-// to the deleted lock.
-func (r *RedisDB) Unlock(project models.Project, workspace string) (*models.ProjectLock, error) {
+// to the deleted lock. If updateQueue is true, it will also grant the
+// lock to the next PR in the queue, update the queue and return the dequeued lock.
+func (r *RedisDB) Unlock(project models.Project, workspace string, updateQueue bool) (*models.ProjectLock, *models.ProjectLock, error) {
 	var lock models.ProjectLock
 	key := r.lockKey(project, workspace)
 
 	val, err := r.client.Get(ctx, key).Result()
 	if err == redis.Nil {
-		return nil, nil
+		return nil, nil, nil
 	} else if err != nil {
-		return nil, errors.Wrap(err, "db transaction failed")
+		return nil, nil, errors.Wrap(err, "db transaction failed")
 	} else {
 		if err := json.Unmarshal([]byte(val), &lock); err != nil {
-			return nil, errors.Wrap(err, "failed to deserialize current lock")
+			return nil, nil, errors.Wrap(err, "failed to deserialize current lock")
 		}
 		r.client.Del(ctx, key)
-		return &lock, nil
+		// Dequeue next item
+		if updateQueue {
```

Comment:
Please break down additions to this function into smaller pieces for more granular unit tests

Reply:
Moved in c04c4ce, but same issue as in https://github.com/runatlantis/atlantis/pull/2782#discussion_r1181297330